host/entry.go

```diff
@@ -0,0 +1,293 @@
+	// name for the host.
+	Name() string
+	// DefaultForm returns the default form.  In case one form is not
+	// implemented or uniform accross all supported functionality (input, output, device),
+	// output default should be returned.
+	DefaultForm() sound.Form
+
+	// DefaultSampleCodec returns the default sample codec.  In case one sample codec
+	// is not implemented or uniform accross all supported functionality, output default
+	// should be returned.
+	DefaultSampleCodec() sample.Codec
+
+	// DefaultBufSize returns the default size of the buffer provided to
+	// the caller for interactions.
+	DefaultBufSize() int
+
+	// CanOpenSource returns true if OpenSource does not return
+	// ErrUnspported.
+	CanOpenSource() bool
+
+	// OpenSource starts capturing audio.
+	//
+	// dev is the specified device, and should be nil if and only if the current
+	// entry has .DevScanner nil.  Otherwise, dev should be a device returned by
+	// the current entry .DevScan()
+	//
+	// v indicates the desired form (channels, sample rate) of the source.
+	//
+	// sco indicates the desired sample.Codec.
+	//
+	// bufSz indicates the size of buffer, in frames, which data from the
+	// hardware or software implementing the entry to sound.Source.Receive.
+	// For hardware, this is normally the size of the part of the ring buffer exposed for
+	// reading.  Implementations should use a minimal total buffer size to accomodate
+	// this constraint.
+	//
+	// OpenSource returns a triple (s, t, e) with
+	// s: sound.Source which represents captured audio.
+	// t: the start time of the first sample.
+	// e: any error
+	OpenSource(dev *libsio.Dev, v sound.Form, sco sample.Codec, bufSz int) (sound.Source, time.Time, error)
+
+	// Can open sink returns true if OpenSink does not return
+	// ErrUnsupported
+	CanOpenSink() bool
+	// OpenSink starts playing audio.
+	//
+	// dev is the specified device, and should be nil if and only if the current
+	// entry has .DevScanner nil.  Otherwise, dev should be a device returned by
+	// the current entry .DevScan()
+	//
+	// v indicates the desired form (channels, sample rate) of the source.
+	//
+	// sco indicates the desired sample.Codec.
+	//
+	// bufSz indicates the size of buffer whose data is placed
+	// in sound.Source.Receive.  This is normally the size of the part of
+	// the ring buffer exposed for playback.  Implementations should
+	// use a minimal total buffer size to safely accomodate this constraint.
+	//
+	// OpenSink returns a tuple (s, *t, e) with
+	// s: sound.Sink which represents audio for playback.
+	// t: a pointer to the start time of the first played sample, which is set
+	//    after the first successful send via the returned sound.Sink.
+	// e: any error
+	//
+	// OpenSink should
+	OpenSink(dev *libsio.Dev, v sound.Form, sco sample.Codec, bufSz int) (sound.Sink, *time.Time, error)
+
+	// CanOpenDuplex returns true if OpenDuplex does not return ErrUnsupported.
+	CanOpenDuplex() bool
+	// OpenDuplex
```

Comment:
OpenDuplex still needs to be documented.

Reply:
Is the returned time the start time of audio playback or of audio capture? If audio playback, is a pointer required?

Reply:
Yes, for playback, the start time is a pointer because it is only set in the future after the first successful send of samples.  For capture is not a pointer as receives get data that is already captured.  Duplex would not use a pointer and would return the capture time.  But Maybe Duplex could return both a time pointer for playback and a time for capture start time, then measuring latency would be easier?